### DISCUSSION TITLE
How to run t5 on a GPU of a local machine?

### DISCUSSION THREAD

**Discussion by puraminy** [google-research/text-to-text-transfer-transformer#785] [Q&A]
I'm in Iran and unfortunately many google services including Google Cloud are banned in Iran. Even if it wasn't banned, we have no international Credit Cards. So, we should use proxies to use these services.

I want to test T5 on my GPU. The command-line instruction apparently requires some gs://* config files and pretrained models.
For example, I founded splitted checkpoints at https://console.cloud.google.com/storage/browser/t5-data/pretrained_models/small;

I tried to copy them on my pc using:

`gsutil cp gs://t5-data/pretrained_models/small/model.ckpt-1000000* .`

It's total of 18 objects with 166Mb
Okay!
Now should I merge them into one file? If yes, with which command?

Thank you

**Comment by craffel** [comment]
Sorry about the access issues. No, they are meant to stay as separate files.

**Reply by puraminy** [reply]
Thank you, please tell me if it can be run locally (without cloud)?
I found no or few hints on how to program and run it on a local GPU.

Anyways, I tried to copy the notebook example and adapt it for my data. In the following, you can see my changes to set up the model, and using local directories! 
I also copied the contents of the google cloud bucket for pretrained models on a local folder. I also set `tpu` to `None` because it was a required parameter.

Are they okay? 

```
MODEL_SIZE = "small"
MODELS_DIR = "/drive3/pouramini/models/t5/"
PRETRAINED_DIR = "/home/pouramini/pretrained_models/small/"

MODEL_DIR = os.path.join(MODELS_DIR, MODEL_SIZE)

# Set parallelism and batch size to fit on v2-8 TPU (if possible).
# Limit number of checkpoints to fit within 5GB (if possible).
model_parallelism, train_batch_size, keep_checkpoint_max = {
    "small": (1, 16, 16),
    "base": (2, 8, 8),
    "large": (8, 1, 4),
    "3B": (8, 16, 1),
    "11B": (8, 16, 1)}[MODEL_SIZE]

tf.io.gfile.makedirs(MODEL_DIR)
# The models from our paper are based on the Mesh Tensorflow Transformer.
model = t5.models.MtfModel(
    tpu=None,
    model_dir=MODEL_DIR,
    batch_size=train_batch_size,
    sequence_length={"inputs": 128, "targets": 32},
    mesh_shape = 'model:1,batch:1',
    mesh_devices = ['gpu:0'],
    learning_rate_schedule=0.003,
    save_checkpoints_steps=5000,
    keep_checkpoint_max=None, #keep_checkpoint_max if ON_CLOUD else None,
    iterations_per_loop=100,
)

model.finetune(
    mixture_or_task_name="nq_context_free",
    pretrained_model_dir=PRETRAINED_DIR,
    finetune_steps=FINETUNE_STEPS
)

```

However, when I run it, I got stuck to an internet authentication:

`tensorflow/core/platform/cloud/google_auth_provider.cc:178] All attempts to get a Google authentication bearer token failed, returning an empty token. Retrieving token from files failed with "Not found: Could not locate the credentials file." ....`

If I set proxy it fails because of timeout!

I don't know where it wants to establish a connection and why? It's related to this:

`	 when reading metadata of gs://t5-data/vocabs/cc_all.32000/sentencepiece.model`

is there a setting to change it?

**Reply by puraminy** [reply]
okay, I got it into running by this code:

```
import t5.data.utils as ut

ut.DEFAULT_SPM_PATH = "/home/pouramini/gins/sentencepiece.model"

```
However, I don't know if there were better solutions to pass the vocabulary and vocabulary path.
